mne/beamformer/_rap_music.py

```diff
@@ -0,0 +1,241 @@
+        info, forward, label=None, picks=picks, pick_ori=None)
+
+    gain = G.copy()
+
+    # Handle whitening + data covariance
+    whitener, _ = compute_whitener(noise_cov, info, picks)
+    if info['projs']:
+        whitener = np.dot(whitener, proj)
+
+    # whiten the leadfield and the data
+    G = np.dot(whitener, G)
+    data = np.dot(whitener, data)
+
+    eig_values, eig_vectors = linalg.eigh(np.dot(data, data.T))
+    phi_sig = eig_vectors[:, -n_dipoles:]
+
+    n_orient = 3 if is_free_ori else 1
+    n_channels = G.shape[0]
+    A = np.empty((n_channels, n_dipoles))
+    gain_dip = np.empty((n_channels, n_dipoles))
+    oris = np.empty((n_dipoles, 3))
+    poss = np.empty((n_dipoles, 3))
+
+    G_proj = G.copy()
+    phi_sig_proj = phi_sig.copy()
+
+    for k in range(n_dipoles):
+        subcorr_max = -1.
+        for i_source in range(G.shape[1] // n_orient):
+            idx_k = slice(n_orient * i_source, n_orient * (i_source + 1))
+            Gk = G_proj[:, idx_k]
+            if n_orient == 3:
+                Gk = np.dot(Gk, forward['source_nn'][idx_k])
+
+            subcorr, ori = _compute_subcorr(Gk, phi_sig_proj)
+            if subcorr > subcorr_max:
+                subcorr_max = subcorr
+                source_idx = i_source
+                source_ori = ori
+                if n_orient == 3 and source_ori[-1] < 0:
+                    # make sure ori is relative to surface ori
+                    source_ori *= -1  # XXX
+
+                source_pos = forward['source_rr'][i_source]
+                if n_orient == 1:
+                    source_ori = forward['source_nn'][i_source]
+
+        idx_k = slice(n_orient * source_idx, n_orient * (source_idx + 1))
+        Ak = G[:, idx_k]
+        if n_orient == 3:
+            Ak = np.dot(Ak, np.dot(forward['source_nn'][idx_k], source_ori))
+
+        A[:, k] = Ak.ravel()
+
+        if return_explained_data:
+            gain_k = gain[:, idx_k]
+            if n_orient == 3:
+                gain_k = np.dot(gain_k,
+                                np.dot(forward['source_nn'][idx_k],
+                                       source_ori))
+            gain_dip[:, k] = gain_k.ravel()
+
+        oris[k] = source_ori
+        poss[k] = source_pos
+
+        logger.info("source %s found: p = %s" % (k + 1, source_idx))
+        if n_orient == 3:
+            logger.info("ori = %s %s %s" % tuple(oris[k]))
+
+        projection = _compute_proj(A[:, :k + 1])
+        G_proj = np.dot(projection, G)
+        phi_sig_proj = np.dot(projection, phi_sig)
+
+    sol = linalg.lstsq(A, data)[0]
+
+    gof, explained_data = [], None
+    if return_explained_data:
+        explained_data = np.dot(gain_dip, sol)
+        gof = (linalg.norm(np.dot(whitener, explained_data)) /
+               linalg.norm(data))
+
+    return _make_dipoles(times, forward['src'], poss,
+                         oris, sol, gof), explained_data
+
+
+def _make_dipoles(times, src, poss, oris, sol, gof):
```

Comment:
can you add a docstring describing the parameters so that the next person has an easier time :)